server/src/main/java/org/elasticsearch/common/time/JavaDateFormatter.java

```diff
@@ -79,36 +77,31 @@ private JavaDateFormatter(String format, DateTimeFormatter printer, DateTimeForm
         }
         this.printer = printer;
         this.format = format;
+
         if (parsers.length == 0) {
-            this.parser = printer;
-        } else if (parsers.length == 1) {
-            this.parser = parsers[0];
+            this.parsers = Collections.singletonList(printer);
         } else {
-            DateTimeFormatterBuilder builder = new DateTimeFormatterBuilder();
-            for (DateTimeFormatter parser : parsers) {
-                builder.appendOptional(parser);
-            }
-            this.parser = builder.toFormatter(Locale.ROOT);
+            this.parsers = Arrays.asList(parsers);
         }
 
         DateTimeFormatterBuilder builder = new DateTimeFormatterBuilder();
         if (format.contains("||") == false) {
-            builder.append(this.parser);
+            builder.append(firstParser());
         }
         roundupParserConsumer.accept(builder);
-        DateTimeFormatter roundupFormatter = builder.toFormatter(parser.getLocale());
+        DateTimeFormatter roundupFormatter = builder.toFormatter(firstParser().getLocale());
         if (printer.getZone() != null) {
             roundupFormatter = roundupFormatter.withZone(printer.getZone());
         }
         this.roundupParser = roundupFormatter;
     }
 
-    DateTimeFormatter getRoundupParser() {
-        return roundupParser;
+    private DateTimeFormatter firstParser() {
```

Comment:
Personal opinion: I dislike this method somehow. There are alternatives whenever this method is called (using `zone()` or `locale()` directly, and calling `this.parsers.get(0)` directly for the rest.

Reply:
my intention was to avoid having repetitive calls of `this.parsers.get(0)`. 
But I see your point that we could use alternatives. Does this mean that `zone()` and `locale()` is always guaranteed  to be the same across `printer` and all `parsers`?

Reply:
see the ctor check for distinct locale and distinct zone